### PR TITLE
Add AgGroup command

### PR DESCRIPTION
**Pull request by albfan**
Allow using Ag as it behaves on shell, but adding the powerful of vim (fold, highlight, search, movement...)

I have added a small screencast for demonstration. Seems speed is a little to fast

Enhancements:
- [x] Use preview window instead of a new buffer
- [x] Align line and column numbers
- [ ] Apply syntax highlight per filetype. Use same magic as [xolox/vim-notes code highlight](https://github.com/xolox/vim-notes/blob/d30601243d989c8df11956f0637106d7f255a051/autoload/xolox/notes.vim#L1160)

### CODE REVIEW

**Comment by losingkeys** [comment]
Thanks for the pull! I'll comment inline when I have a chance to review.

**Comment by albfan** [comment]
Hey,

although labels are self-explanatory, you might open a wiki with the workflow on ag.vim. That way you can ease the PR process.

I have added some test to AgGroup using [Vader](https://github.com/junegunn/vader.vim) so you can use a [TDD](http://en.wikipedia.org/wiki/Test-driven_development) to check this and future changes.

I'm checking results and syntax

**Comment by Tuch** [comment]
Best request

**Comment by losingkeys** [comment]
Nice work. Trying it out locally, it's pretty slick; I like the folds, especially because I have space mapped to `zMzjzo`.

Some things I noticed:
- The keybindings don't work in this window (I think they'd have to be buffer-local to avoid mapping things in other scratch buffers). Also once they do work we'll need the 'help text' that shows up after other searches.
- `:AgGroup` without any terms shows the following error:
```
:AgGroup                                                                                     
shell returned 1                                                                             
Error detected while processing function ag#AgGroup:                                         
line   22:                                                                                   
E684: list index out of range: 0                                                             
E15: Invalid expression: split(a:args, '\s\+')[0]                                            
line   25:                                                                                   
E475: Invalid argument: agSearch //                                                          
Press ENTER or type command to continue                                                      
```
- The enter behavior is different than the normal quickfix/location list behavior. When you search with `:Ag`, then hit enter to jump to a match, it re-uses a window. `:AgGroup` opens new windows always

Thanks again for the pull. I'll try to find more time to do an actual code review too :).

**Comment by albfan** [comment]
I'll check and correct all that unwanted behaviours. The reuse of window is one of the most remarkable gap, as it works on `:Ag`. Let me review it and fix it.

**Comment by albfan** [comment]
Now AgGroup is reusing previewwindow. I haven't define a multiline errformat so I guess quickfix window is not the place to show search with this implementation

I have adopted space as quick navigator (it is a stylish solution, thanks)

Even more it tries to locate the right window to open selection

Now I will refine error handling and daily usage (search visual selection, hide line numbers or expand context without lossing actual position)

**Comment by Numkil** [comment]
@albfan really cool stuff. I like it. Could you add some comments to the big agGroup function? You don't need to rewrite the vim manual but just describe the flow of what's going on, and clarify some potentially complex parts.

File: autoload/ag.vim [hunk 1]
```diff
@@ -58,6 +58,112 @@ function! ag#AgBuffer(cmd, args)
...
+
+function! ToggleEntireFold()
+  if foldclosed(2) == -1
+    normal zM
+  else
+    normal zR
+  endif
+endfunction
+
+function! FoldAg()
```
Comment (Numkil) [inline autoload/ag.vim]: You should be able to rewrite this to
```vimscript
   let line = getline(v:lnum)
   if !empty(line)
      return '1'
   endif
   return '0'
endfunction
```

File: autoload/ag.vim [hunk 2]
```diff
@@ -58,6 +58,112 @@ function! ag#AgBuffer(cmd, args)
...
+  highlight link agLine LineNr
+  highlight link agFile Question
+  highlight link agSearch Todo
+  highlight link agLineContext Constant
+  setlocal foldmethod=expr
+  setlocal foldexpr=FoldAg()
+  setlocal foldcolumn=2
+  1
+  setlocal nomodifiable
+  map <buffer> o za
```
Comment (Numkil) [inline autoload/ag.vim]: It's probably a good idea to use 
```viml
nnoremap <buffer> <silent> 
```
Just to be consistent with the rest of the plugin.

Comment (Numkil) [inline autoload/ag.vim]: Tbh I would also expect this buffer to have the same mappings as the other buffer's created by ag.vim. It's kinda confusing that open vertical split and  the like are not available, and that o doesn't do the same in this buffer than in the others. What do you think? Would it be a good idea to reuse the existing mappings in this command?

**Comment by albfan** [comment]
After continuing with those cleanup and standarization I wanted to focus on selection. Although visual selection leads to a lot of use cases erroneous `:AgGroup function name(` this smart selection it works pretty well

```viml
map <Leader>ag :<C-U>AgGroup<CR>
```

test it on normal or visual mode. it rocks!

**Comment by amerlyq** [comment]
@rking, please, merge it already. I reviewed all commits and even contradictory changes (like forced smartcase) are perfectly reasonable for common usecase. If postponed more, it will only become harder to merge later. Any other changes/improvements to this pull request are better to be separated. Because of current forks diversity I can't maintain my own pull request changing both forks code -- and so it psychologically exhausts me.

**Comment by albfan** [comment]
I think best approach at this moment is to close this PR and rebase all code. Then start to cherry pick approved changes from discussion on new opened issues. Let's do a divide an conquer. 

Here my fault was to PR from master, definitely real bad workflow. I will cover fork features with test to be sure they work as expected, let me know if you are interested on add test suite to ag.vim

**Comment by albfan** [comment]
I will try to collaborate with @amerlyq in revamp this PR to ag.vim tip and split int small pieces ready to merge on upstream

**Pull request closed without merging**